### ISSUE TITLE
Performance issues on huge documents

### ISSUE THREAD

**Issue by sowasred2012** [ianstormtaylor/slate#1267]
#### Do you want to request a *feature* or report a *bug*?

Bug.

#### What's the current behavior?

Huge documents are currently very slow in dealing with new text input. I don't have a JSFiddle, but the issue can be seen on [the demo site](http://slatejs.org/#/huge-document).

I'm using OSX 10.12.4, Chrome 61 and I imagine the demo site is using the latest version of Slate.

Gif:

![lag2](https://user-images.githubusercontent.com/2376968/31813216-7aa742c0-b57d-11e7-9507-c931e17ccc78.gif)

Dev Tools performance profile:

![screen shot 2017-10-20 at 09 42 30](https://user-images.githubusercontent.com/2376968/31813235-8dacf27a-b57d-11e7-82dc-8ce898393f8a.png)

I've saved this profile, happy to share it if need be (42mb).

#### What's the expected behavior?

Text input should be a lot more responsive. Would love to suggest a solution but I'm relatively new to React! Worth noting it's related to an older issue that was closed: #119

**Comment by doodlewind** [comment]
Performance is dramatically improved by #1300. Seems that Slate is usable when editing huge documents now. @ianstormtaylor great job!

**Comment by sowasred2012** [comment]
What @doodlewind said! Huge doc works great now :)

**Comment by bogdanbrezuica** [comment]
Sorry to bring bad news, but it seems this problem is still reproduced on the "huge document" slate demo. It works ok when typing, but when pressing enter the editor freezes for a few seconds (~3.5s in my case). Probably unrelated, but I started investigating performance issues when I noticed the "history" property on Value is growing exponentially with each change made to the document. After a few changes, when I try to inspect the Value object `value.toJS()` the app freezes because the object is huge (because of the history property).

@ianstormtaylor any ideas? btw, I really appreciate your work on Slate!

This is a performance profile for the "huge document" demo on slatejs.org
![slate performance profile](https://user-images.githubusercontent.com/13481438/37082951-99d6a5e2-21f6-11e8-9425-9bdac32f60b4.PNG)
